source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include "source/extensions/config_subscription/grpc/grpc_stream.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * This class arbitrates between two config providers of the same GrpcMux -
+ * the primary and the failover. Envoy always prefers fetching config from the
+ * primary source, but if not available, will fetch the config from the failover
+ * source until the primary is again available.
+ *
+ * This class owns the state for the GrpcMux primary and failover streams, and
+ * proxies the gRPC stream functionality to either the primary or the failover config sources.
+ * The failover source is optional and will only be used if passed upon initialization.
+ *
+ * The primary config source is always preferred over the failover source. If the primary
+ * is available, the stream to the failover source will terminated.
+ *
+ * Failover is supported in both
+ * SotW (RequestType = envoy::service::discovery::v3::DiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DiscoveryResponse), and
+ * Delta-xDS (RequestType = envoy::service::discovery::v3::DeltaDiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DeltaDiscoveryResponse).
+ * Both the primary and failover streams are either SotW or Delta-xDS.
+ *
+ * The use of this class will be as follows: the GrpcMux object will own an instance of
+ * the GrpcMuxFailoverManager. The GrpcMuxFailoverManager will own 2 GrpcStreams, primary
+ * and secondary. Each of the primary and secondary streams will invoke GrpcStreamCallbacks
+ * on their corresponding objects (also owned by the GrpcMuxFailoverProxy). These invocations
+ * will be followed by the GrpcMuxFailoverProxy calling the GrpcStreamCallbacks on the GrpcMux
+ * object that initialized it.
+ *
+ * Note: this class is WIP.
+ */
+template <class RequestType, class ResponseType> class GrpcMuxFailover {
+public:
+  // A GrpcStream creator function that receives the stream callbacks and returns a
+  // GrpcStream object. This is introduced to facilitate dependency injection for
+  // testing and will be used to create the primary and failover streams.
+  using GrpcStreamCreator = std::function<GrpcStreamInterfacePtr<RequestType, ResponseType>(
+      GrpcStreamCallbacks<ResponseType>* stream_callbacks)>;
+
+  GrpcMuxFailover(GrpcStreamCreator primary_stream_creator,
+                  OptRef<GrpcStreamCreator> failover_stream_creator,
+                  GrpcStreamCallbacks<ResponseType>& grpc_mux_callbacks)
+      : currently_used_service_(ServiceOptions::None), grpc_mux_callbacks_(grpc_mux_callbacks),
+        primary_callbacks_(*this),
```

Comment:
why do we need to store this in primary_callbacks_, can we just pass `this` to primary_stream_creator` below?

Reply:
Not sure I fully understand the suggestion, so I'll try to give some more info:
The GrpcStream needs to know which callbacks to invoke.
Currently the GrpcMux owns the GrpcStream, and when initializing it, it passes "this" for callback invocation.
In this PR we add the GrpcMuxFailover, that will (in future PRs) be between the GrpcMux and 2 GrpcStream objects, for primary and failover.

There are going to be 2 types of callback helpers:
`PrimaryGrpcStreamCallbacksWrapper` (defined in this file below) will process the callbacks from the primary source, and ensure that the GrpcxMuxFailover is aware that the primary source is available.
`FailoverGrpcStreamCallbacksWrapper` (to be introduced later) will process the callbacks for the failover source, and ensure that these callbacks are only invoked back at the GrpcMux if the primary isn't available.

Reply:
Got it, I understand the separation in callbacks now, thanks for the explanation!